config/v1/types_cluster_version.go

```diff
@@ -106,6 +111,48 @@ type ClusterVersionStatus struct {
 	AvailableUpdates []Update `json:"availableUpdates"`
 }
 
+// UpdateState is a constant representing whether an update was successfully
+// applied to the cluster or not.
+type UpdateState string
+
+const (
+	// CompletedUpdate indicates an update was successfully applied
+	// to the cluster (all resource updates were successful).
+	CompletedUpdate UpdateState = "Completed"
+	// PartialUpdate indicates an update was never completely applied
+	// or is currently being applied.
+	PartialUpdate UpdateState = "Partial"
+)
+
+// UpdateHistory is a single attempted update to the cluster.
+type UpdateHistory struct {
+	// state reflects whether the update was fully applied. The Partial state
+	// indicates the update is not fully applied, while the Completed state
+	// indicates the update was successfully rolled out at least once (all
+	// parts of the update successfully applied).
+	State UpdateState `json:"state"`
+
+	// startedTime is the time at which the update was started.
+	StartedTime metav1.Time `json:"startedTime"`
+	// completionTime, if set, is when the update was fully applied. The update
+	// that is currently being applied will have a null completion time.
+	// Completion time will always be set for entries that are not the current
+	// update (usually to the started time of the next update).
+	CompletionTime *metav1.Time `json:"completionTime"`
+
+	// version is a semantic versioning identifying the update version. When this
+	// field is part of spec, version is optional if payload is specified.
```

Comment:
I'd like to have an indication in the payload of what version it is.  I don't think it should ever be optional here.

Reply:
It's optional if the payload doesn't have a version.  Some payloads won't have a version.  Also, the CVO may not know what the payload version is before it is requested (you can provide only an image and we have to honor it, and if you provide a false version we have to distinguish that as well).

In general, yes, we will have a version, but we need to be clear you can't always get one.